src/Mvc/Mvc.ViewFeatures/src/Infrastructure/DefaultTempDataSerializer.cs

```diff
@@ -150,7 +232,10 @@ public override bool CanSerializeType(Type type)
                 type == typeof(string) ||
                 type == typeof(bool) ||
                 type == typeof(DateTime) ||
-                type == typeof(Guid);
+                type == typeof(Guid) ||
+                typeof(ICollection<int>).IsAssignableFrom(type) ||
+                typeof(ICollection<string>).IsAssignableFrom(type) ||
+                typeof(IDictionary<string, string>).IsAssignableFrom(type);
```

Comment:
Json.Net supports a bunch more collection types like `List<DateTime>, and `Dictionary<string, Guid>`, but we have to write our own dictionary deserialization for all of these scenarios with System.Text.Json. I'm limiting it to a couple of meaningful collection types. We can always tell users to continue using Json.NET if they run in to issues.

Reply:
Or even better they can use strings 😆

Reply:
I don't really want to set up us to add a million different options here. like, I don't think it's a success condition if users have to ask to rebuild the JSON serializer piecemeal. 

Ultimately we're going to get support for the serializer to read from the reader directly. Does that solve this issue once and for all for us or will there still be limitations?

Reply:
> Does that solve this issue once and for all for us or will there still be limitations?

In some ways. The larger problem is that we have to inspect the value and determine what the value intended to stash in TempData`. We essentially need a limited form of `TypeNameHandling`

Reply:
Right. I remember the issue now. Exposing TempData as `IDictionary<string, object>` is just fundamentally problematic. The API contract implies that we will round-trip values with their original runtime type.